x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/build_exceptions_query.ts

```diff
@@ -0,0 +1,151 @@
+    return value.toUpperCase();
+  }
+
+  return value;
+};
+
+export const operatorBuilder = (operator: ListOperator, language: string): string => {
+  return operator === 'excluded'
+    ? ` ${getLanguageBooleanOperator(language, 'and')} `
+    : ` ${getLanguageBooleanOperator(language, 'and')} ${getLanguageBooleanOperator(
+        language,
+        'not'
+      )} `;
+};
+
+export const buildExists = ({
+  operator,
+  field,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  language: string;
+}): string => {
+  switch (language) {
+    case 'kuery':
+      return `${operatorBuilder(operator, language)}${field}:*`;
+    case 'lucene':
+      return `${operatorBuilder(operator, language)}_exists_${field}`;
+    default:
+      return '';
+  }
+};
+
+export const buildMatch = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  const value = values[0].name;
+  return `${operatorBuilder(operator, language)}${field}:${value}`;
+};
+
+export const buildMatchAll = ({
+  operator,
+  field,
+  values,
+  language,
+}: {
+  operator: ListOperator;
+  field: string;
+  values: ListValues[];
+  language: string;
+}) => {
+  if (values.length === 1) {
+    return buildMatch({ operator, field, values, language });
+  } else {
+    const matchAllValues = values.map(value => {
+      return value.name;
+    });
+
+    return `${operatorBuilder(operator, language)}${field}:(${matchAllValues.join(
+      ` ${getLanguageBooleanOperator(language, 'or')} `
+    )})`;
+  }
+};
+
+export const evaluateValues = ({ list, language }: { list: List; language: string }): string => {
+  const { values_operator: operator, values_type: type, field, values } = list;
+  switch (type) {
+    case 'exists':
+      return buildExists({ operator, field, language });
+    case 'match':
+      return buildMatch({ operator, field, values: values!, language });
+    case 'match_all':
+      return buildMatchAll({ operator, field, values: values!, language });
+    default:
+      return '';
+  }
+};
+
+export const buildExceptions = ({
+  query,
+  lists,
+  language,
+  accumulator = [],
+}: {
+  query: string;
+  lists: List[];
+  language: string;
+  accumulator?: string[];
+}): string[] => {
+  return lists.reduce((acc, item) => {
+    const exception: string[] = [query];
+    const { and, ...exceptionDetails } = { ...item };
+
+    exception.push(evaluateValues({ list: exceptionDetails, language }));
+
+    if (and && !isEmpty(and)) {
+      and.forEach(i => {
+        exception.push(evaluateValues({ list: i, language }));
+      });
+    }
+
+    acc.push(`(${exception.join('')})`);
+    return acc;
+  }, accumulator);
+};
+
+export const buildQueryExceptions = ({
+  query,
+  language,
+  lists,
+}: {
+  query: string;
+  language: string;
+  lists?: RuleAlertParams['lists'];
+}): Query[] => {
+  if (lists && lists !== null && !isEmpty(lists)) {
```

Comment:
This is a lot of overlapping boolean logic checking here?

You can just do:

```ts
if (lists != null && !isEmpty(lists)) {
```

And that covers, `undefined` and `null`. I rarely use `!isEmpty()` from lodash anymore fwiw as empty arrays usually and should just work and also lodash isEmpty no longer connects well with TypeScript. TypeScript understands the `!= null` pretty well but does not look at the `!isEmpty(lists)` and understand that it is implying that theres redundant undefined and null checks.

The less lodash I use these days the better off I typically feel as it's one less framework to worry about in the codebase. But that part is optional. It looks like you want to explicitly check for not an empty array here in this case.

Reply:
I'll admit that the overlap was to appease TypeScript - which I know is not great! It didn't seem to acknowledge that `isEmpty` would catch the undefined. I switched it up and removed the use of lodash so that this function deals with the logic of `buildExceptions` returning back (and accepting an empty array for lists) an empty array. I also added tests to explicitly test when lists is an empty array, null and undefined. Thanks for pointing this out!